File: packages/ogdf.js/src/module/src/CrossingMinimization.cpp
#include "../main.h"

#include <ogdf/planarity/SubgraphPlanarizer.h>

EM_PORT_API(SubgraphPlanarizer *)
CrossingMinimizationModule_SubgraphPlanarizer(int globalInternalLibraryLogLevel, int globalLogLevel, int globalMinimumLogLevel, bool globalStatisticMode, int localLogLevel, int localLogMode, unsigned int maxThreads, int permutations, bool timeout, double timeLimit, EdgeInsertionModule *inserter, PlanarSubgraphModule<int> *subgraph)
{
    SubgraphPlanarizer *subgraphPlanarizer = new SubgraphPlanarizer();
    // subgraphPlanarizer->globalInternalLibraryLogLevel(static_cast<Logger::Level>(globalInternalLibraryLogLevel));
    // subgraphPlanarizer->globalLogLevel(static_cast<Logger::Level>(globalLogLevel));
    // subgraphPlanarizer->globalMinimumLogLevel(static_cast<Logger::Level>(globalMinimumLogLevel));
    // subgraphPlanarizer->globalStatisticMode(globalStatisticMode);
    // subgraphPlanarizer->localLogLevel(static_cast<Logger::Level>(localLogLevel));
    // subgraphPlanarizer->localLogMode(static_cast<Logger::LogMode>(localLogMode));
    subgraphPlanarizer->maxThreads(maxThreads);
    subgraphPlanarizer->permutations(permutations);
    subgraphPlanarizer->setTimeout(timeout);
    subgraphPlanarizer->timeLimit(timeLimit);
    subgraphPlanarizer->setInserter(inserter);
    subgraphPlanarizer->setSubgraph(subgraph);

    return subgraphPlanarizer;
}

File: packages/ogdf.js/src/module/config/CrossingMinimization.js
import EdgeInsertionModule from "./EdgeInsertion";
import createModule from "../module-helper";
import PlanarSubgraphModule from "./PlanarSubgraph";
import { PARAMETER_TYPE } from "../../utils/parameter-type";

const NAME = "CrossingMinimizationModule"

const MODULE_DIRECTORY = {
    SubgraphPlanarizer: {
        globalInternalLibraryLogLevel: {
            type: PARAMETER_TYPE.CATEGORICAL,
            range: ['Minor', 'Medium', 'Default', 'High', 'Alarm', 'Force'],
            default: 'Alarm'
        },
        globalLogLevel: {
            type: PARAMETER_TYPE.CATEGORICAL,
            range: ['Minor', 'Medium', 'Default', 'High', 'Alarm', 'Force'],
            default: 'Alarm'
        },
        globalMinimumLogLevel: {
            type: PARAMETER_TYPE.CATEGORICAL,
            range: ['Minor', 'Medium', 'Default', 'High', 'Alarm', 'Force'],
            default: 'Medium'
        },
        globalStatisticMode: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: false
        },
        localLogLevel: {
            type: PARAMETER_TYPE.CATEGORICAL,
            range: ['Minor', 'Medium', 'Default', 'High', 'Alarm', 'Force'],
            default: 'Default'
        },
        localLogMode: {
            type: PARAMETER_TYPE.CATEGORICAL,
            range: ['Global', 'GlobalLog', 'Log', 'Statistic'],
            default: 'Global'
        },
        maxThreads: {
            type: PARAMETER_TYPE.INT,
            range: [0, Infinity],
            default: 1
        },
        permutations: {
            type: PARAMETER_TYPE.INT,
            range: [0, Infinity],
            default: 1
        },
        timeout: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        },
        timeLimit: {
            type: PARAMETER_TYPE.INT,
            range: [-Infinity, Infinity],
            default: 1000
            // default: -1 if timeLimit < 0 will be unstoppable
        },
        inserter: {
            type: PARAMETER_TYPE.MODULE,
            module: EdgeInsertionModule,
            default: EdgeInsertionModule.VariableEmbeddingInserter
        },
        subgraph: {
            type: PARAMETER_TYPE.MODULE,
            module: PlanarSubgraphModule,
            default: PlanarSubgraphModule.PlanarSubgraphFast
        }
    }
}

const CrossingMinimizationModule = createModule(NAME, MODULE_DIRECTORY)
export default CrossingMinimizationModule